aptos-move/aptos-vm/src/block_executor/mod.rs

```diff
@@ -0,0 +1,190 @@
+            .collect()
+    }
+
+    /// Execution output for transactions that comes after SkipRest signal.
+    fn skip_output() -> Self {
+        Self(TransactionOutputExt::from(TransactionOutput::new(
+            WriteSet::default(),
+            vec![],
+            0,
+            TransactionStatus::Retry,
+        )))
+    }
+}
+
+pub struct BlockAptosVM();
+
+impl BlockAptosVM {
+    fn process_parallel_block_output<S: StateView>(
+        results: Vec<AptosTransactionOutput>,
+        delta_resolver: OutputDeltaResolver<StateKey, WriteOp>,
+        state_view: &S,
+    ) -> Vec<TransactionOutput> {
+        // TODO: MVHashmap, and then delta resolver should track aggregator keys.
+        let mut aggregator_keys: HashMap<StateKey, anyhow::Result<ResolvedData>> = HashMap::new();
+        for res in results.iter() {
+            let output_ext = AptosTransactionOutput::as_ref(res);
+            for (key, _) in output_ext.delta_change_set().iter() {
+                if !aggregator_keys.contains_key(key) {
+                    aggregator_keys.insert(key.clone(), state_view.get_state_value(key));
+                }
+            }
+        }
+
+        let materialized_deltas =
+            delta_resolver.resolve(aggregator_keys.into_iter().collect(), results.len());
+
+        results
+            .into_iter()
+            .zip(materialized_deltas.into_iter())
+            .map(|(res, delta_writes)| {
+                let output_ext = AptosTransactionOutput::into(res);
+                output_ext.output_with_delta_writes(WriteSetMut::new(delta_writes))
+            })
+            .collect()
+    }
+
+    fn process_sequential_block_output(
+        results: Vec<AptosTransactionOutput>,
+    ) -> Vec<TransactionOutput> {
+        results
+            .into_iter()
+            .map(|res| {
+                let output_ext = AptosTransactionOutput::into(res);
+                let (deltas, output) = output_ext.into();
+                debug_assert!(deltas.is_empty(), "[Execution] Deltas must be materialized");
+                output
+            })
+            .collect()
+    }
+
+    pub fn execute_block<S: StateView>(
+        transactions: Vec<Transaction>,
+        state_view: &S,
+        concurrency_level: usize,
+    ) -> Result<Vec<TransactionOutput>, VMStatus> {
+        // Verify the signatures of all the transactions in parallel.
+        // This is time consuming so don't wait and do the checking
+        // sequentially while executing the transactions.
+        let signature_verified_block: Vec<PreprocessedTransaction> =
+            RAYON_EXEC_POOL.install(|| {
+                transactions
+                    .par_iter()
+                    .map(|txn| preprocess_transaction::<AptosVM>(txn.clone()))
+                    .collect()
+            });
+
+        let executor =
+            BlockExecutor::<PreprocessedTransaction, AptosVMWrapper<S>>::new(concurrency_level);
+
+        let mut ret = if concurrency_level > 1 {
```

Comment:
I agree, I think this will be much more natural once I also hide the delta application and fallbacks inside the block_executor crate, then it can have execute_block() and internally deal with the concurrency_level (implementation detail). Now we kinda deal with some of it here, in the wrapper and some in the executor.

The challenges to getting there is simply to unify the output processing for both parallel and sequential modes (much like I am trying to unify the execute_transaction_* callback interfaces) and move it to block_executor. That would involve a bit of massaging of output_delta_resolver, which I can do as yet another separate diff.

Reply:
here we go, draft follow-up to the draft follow-up I linked before:
https://github.com/aptos-labs/aptos-core/commit/d2e7f81f5c05ae4137104697e9cc964223c2a088
basically you can look how much simpler this file (aptos-vm/block_executor/mod.rs) becomes.

The logic moves to block_executor crate. This should make everything even nicer, and additionally can optimize the flow for generating which keys correspond to aggregators (previously required an exhaustive traversal over all transactions all affected keys, don't think it was even done in parallel), now we just filter & collect from MVHashMap keys and Delta resolution becomes an implementation detail.